src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -510,14 +562,180 @@ internal void ReadToken(string encodedJson)
                 }
             }
 
-            EncodedToken = encodedJson;
+            _encodedToken = encodedJson;
+        }
+
+#if NET6_0_OR_GREATER
+        /// <summary>
+        /// Converts a span into an instance of <see cref="JsonWebToken"/>.
+        /// </summary>
+        /// <param name="encodedTokenMemory">A span representing a 'JSON Web Token' (JWT) in JWS or JWE Compact Serialization Format.</param>
+        /// <exception cref="SecurityTokenMalformedException">if <paramref name="encodedTokenMemory"/> is malformed, a valid JWT should have either 2 dots (JWS) or 4 dots (JWE).</exception>
+        /// <exception cref="SecurityTokenMalformedException">if <paramref name="encodedTokenMemory"/> does not have an non-empty authentication tag after the 4th dot for a JWE.</exception>
+        /// <exception cref="SecurityTokenMalformedException">if <paramref name="encodedTokenMemory"/> has more than 4 dots.</exception>
+        internal void ReadToken(ReadOnlyMemory<char> encodedTokenMemory)
+        {
+            // JWT must have 2 dots for JWS or 4 dots for JWE (a.b.c.d.e)
+            ReadOnlySpan<char> encodedTokenSpan = encodedTokenMemory.Span;
+
+            Dot1 = encodedTokenSpan.IndexOf("."); 
+            if (Dot1 == -1 || Dot1 == encodedTokenSpan.Length - 1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14100));
+
+            Dot2 = encodedTokenSpan.Slice(Dot1 + 1).IndexOf(".");
+            if (Dot2 == -1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14120));
+
+            Dot2 = Dot1 + Dot2 + 1;
+            Dot3 = (Dot2 == encodedTokenSpan.Length - 1) ? -1 : encodedTokenSpan.Slice(Dot2 + 1).IndexOf('.');
+
+            if (Dot3 == -1)
+            {
+                // JWS has two dots
+                // JWS: https://www.rfc-editor.org/rfc/rfc7515
+                // Format: https://www.rfc-editor.org/rfc/rfc7515#page-7
+
+                IsSigned = !(Dot2 + 1 == encodedTokenSpan.Length);
+                try
+                {
+                    Header = CreateClaimSet(encodedTokenSpan, 0, Dot1, createHeaderClaimSet: true);
```

Comment:
I question the value of added a general CreateClaimSet, then passing a bool to control which ClaimSet to create.
This does not remove the need for CreateHeaderClaimSet of CreatePayloadClaimSet as those are called, but later.

Reply:
CreateClaimSet was kept to avoid code duplication while renting and returning an ArrayPool.

Reply:
Regarding your comment on removing
[CreateClaimSet](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/74232825dce9009c840daa47b20e784cbcb7df3b/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs#L568C9-L581C10)
method and adding an overload for Base64UrlEncoding.Decode that accepts a ReadyOnlySpan<char> I have implemented the suggest modification and saw a regression in memory allocation. It looks like the location of ArrayPool matters and it is more effective when it is rented within the JsonWebToken class and shows a reduction in memory allocation.